tensorflow/core/grappler/optimizers/meta_optimizer.cc

```diff
@@ -272,7 +272,7 @@ Status MetaOptimizer::OptimizeGraph(Cluster* cluster, const GrapplerItem& item,
   }
 
   std::vector<std::unique_ptr<GraphOptimizer>> optimizers;
-  if (cfg_.optimizers().empty()) {
+  if (cfg_.optimizers().empty() && cfg_.custom_optimizers().empty()) {
```

Comment:
This is not correct. cfg_custom_optimizers() is used by InitializeOptimizers() to initialize the custom optimizers. See line 206.

If cfg_.optimizers() is not empty, it should contain the names of all optimizers, custom or not. See line 193.

Reply:
@rmlarsen, we have custom optimizers that require configuration (TRTOptimizationPass). As of now if we add them to the optimizers list they are not initialized. How should we add single custom configurable optimizer?

Reply:
So you want to run your optimizer and nothing else, is that it? In that case, I suppose you would add your optimizer to custom_optimizers, and turn all the default optimizers off in the config.

Reply:
This is one of the use cases but if I want to control the order of optimizations, I need to use custom_optimizers list. In which case I need to disable all defaults manually. However behavior for optimizers list is inverted, i.e. it disables all default optimizers. This is a non-uniform behavior and add complications of maintainability since it requires embedding of rewrite_options schema, i.e. if new optimizers added to the default list, they will be silently applied before custom optimizer list is consumed. 
Also the current logic, if optimizer is in optimizers list, it overrides defaults else if it is in custom_optimizers list, it is appended to whatever the defaults are and configurable optimizers can only be in custom_optimizers list.

If you insist on logic optimizers override and custom_optimizers append, removing line 199 could be a solution except that it blocks a possibility of multi-pass optimizer, that is same optimizer applied multiple times with different parameters.

Reply:
My concern is that your change breaks current use cases of the form "Run the default optimizers + my custom optimizers in cfg.custom_optimizers().

Reply:
Perhaps that may clear use case a little bit. 
Assume that A-F are current optimizers implemented in TF and have individual flags in rewrite_options protobuf. So when I run tensorflow without any special configuration A,B,C,D,E,F applied to the graph once or twice, depending on optimizer.

Now I implemented an optimizer X that requires some configuration parameters. I want to run X in a special place and possibly multiple times. For example I want to run it in the order A, B, X_1, C, X_2 where X_1 and X_2 are X with two different parameter set. In order to do that I need to disable all default settings in rewrite_options. As you can see this is a tedious process and I write a python function "enable_optimizer" out of this. Then at a later time TF adds new optimizers G and H. Then all the code I have using "enable_optimizer"  starts running  G, H, A, B, X_1, C, X_2 instead of running expected A, B, X_1, C, X_2 and I wonder what changed in my code to lead to this behavior.

With this PR, All I need to do is set the sequence in custom_optimizers list and it will be backwards compatible as long as A, B and C is still there. Currently this is only true if X is the last optimizer and I add A,B and C to optimizers list. @rmlarsen does this make my case clearer?

Thanks,
Sami

Reply:
@rmlarsen,

> My concern is that your change breaks current use cases of the form "Run the default optimizers + my custom optimizers in cfg.custom_optimizers().
That is correct. But I have a feeling that run default + my custom is more fragile than setting optimizers. Most of the time, order of optimizations are important and need to be controlled. We frequently observed that when new optimizers are added to the list of default optimizers, performance of our optimizations dropped or optimization completely failed.

Reply:
Thanks Sami for the detailed explanation. In my opinion, if possible, a cleaner solution could be to remove the `RewriterConfig::optimizer` list, and let all optimizers use the `RewriterConfig::custom_optimizers` list (and maybe rename it back to `RewriterConfig::optimizer`). This way, we only have one list. If the list is empty, we enable the default optimizers (layout, constfold, etc) based on `Toggle` options. If the list is not empty, we ignore any `Toggle` options, and add optimizers based on the the list.

I think Sami's solution in this PR is somehow equivalent to what I described, the only difference is there are two lists.

Reply:
@aaroey, I agree that one list would be even better. Only problem with the single list is that users of the optimize list need to change their code to add configurable_optimizer object. Perhaps with a deprecation message from meta_optimizer this would be the best option.

Reply:
@samikama if your optimizers fail after new optimizers are added upstream, it likely means that they are making incorrect assumptions about the graph or are buggy. I agree with you that we likely need separate lists due to the separate base classes used for such custom versus non-custom optimizers.

Reply:
@rmlarsen is there any existing C++/python helper function that can return a `RewriterConfig` that with all default Grappler optimizers disabled? I think this is very useful at least for testing, see e.g. [tf_cnn_benchmark](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/benchmark_cnn.py#L706).